File: backend/mini-job-service/src/main/java/com/example/minijob/user/service/UserService.java
// [FILE]
// - 목적: User 비즈니스 로직
// - 주요 역할: 회원가입, 로그인 처리
// - 관련 토이 버전: [BE-v0.6]
// - 권장 읽는 순서: signup() → login()
//
// [LEARN] 비밀번호는 BCryptPasswordEncoder로 해시한다.
//         로그인 성공 시 JWT 토큰을 발급한다.

package com.example.minijob.user.service;

import com.example.minijob.config.JwtTokenProvider;
import com.example.minijob.user.domain.User;
import com.example.minijob.user.repository.UserRepository;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

// [Order 1] User 서비스
// - 토이 버전: [BE-v0.6]+
// [LEARN] 인증 로직은 서비스 레이어에서 처리한다.
@Service
@Transactional(readOnly = true)
public class UserService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenProvider jwtTokenProvider;

    public UserService(UserRepository userRepository,
                       PasswordEncoder passwordEncoder,
                       JwtTokenProvider jwtTokenProvider) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
        this.jwtTokenProvider = jwtTokenProvider;
    }

    // [Order 2] 회원가입
    // [LEARN] 비밀번호를 해시하여 저장한다.
    @Transactional
    public User signup(String email, String password) {
        if (userRepository.existsByEmail(email)) {
            throw new IllegalArgumentException("Email already exists: " + email);
        }

        String encodedPassword = passwordEncoder.encode(password);
        User user = new User(email, encodedPassword, "USER");
        return userRepository.save(user);
    }

    // [Order 3] 로그인
    // - 이메일/비밀번호 확인 후 JWT 토큰 발급
    // [LEARN] BCrypt의 matches 메서드로 비밀번호를 검증한다.
    public String login(String email, String password) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new IllegalArgumentException("Invalid email or password"));

        if (!passwordEncoder.matches(password, user.getPassword())) {
            throw new IllegalArgumentException("Invalid email or password");
        }

        return jwtTokenProvider.createToken(user.getEmail(), user.getRole());
    }

    // [Order 4] 이메일로 사용자 조회
    public User findByEmail(String email) {
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + email));
    }
}

File: backend/mini-job-service-node/src/user/userService.ts
// [FILE]
// - 목적: 인증/인가 비즈니스 로직
// - 주요 역할: 회원가입, 로그인, JWT 발급
// - 관련 토이 버전: [Node-BE-v0.1]
//
// [LEARN] bcrypt로 비밀번호를 해싱하고, jsonwebtoken으로 stateless 인증을 구성한다.
import bcrypt from 'bcryptjs';
import jwt from 'jsonwebtoken';
import { HttpError } from '../common/errorHandler';
import { appConfig } from '../config/appConfig';
import { UserRepository } from './userRepository';
import { User } from './user';

export class UserService {
  constructor(private readonly repo: UserRepository) {}

  async signup(email: string, password: string): Promise<string> {
    this.validate(email, password);
    if (this.repo.findByEmail(email)) {
      throw new HttpError(409, 'User already exists');
    }
    const hash = await bcrypt.hash(password, 10);
    const user = this.repo.save(email, hash);
    return this.issueToken(user);
  }

  async login(email: string, password: string): Promise<string> {
    this.validate(email, password);
    const existing = this.repo.findByEmail(email);
    if (!existing) {
      throw new HttpError(401, 'Invalid credentials');
    }
    const ok = await bcrypt.compare(password, existing.passwordHash);
    if (!ok) {
      throw new HttpError(401, 'Invalid credentials');
    }
    return this.issueToken(existing);
  }

  private issueToken(user: User): string {
    return jwt.sign({ sub: user.id, email: user.email }, appConfig.jwtSecret, {
      expiresIn: appConfig.jwtExpirationMs / 1000,
    });
  }

  private validate(email: string, password: string) {
    if (!email || !password) {
      throw new HttpError(400, 'email and password are required');
    }
  }
}
